crypto/signatures_test.go

```diff
@@ -1,25 +1,141 @@
 package crypto
 
 import (
+	"bytes"
 	"crypto/rand"
+	"errors"
 	"testing"
 
 	"github.com/NebulousLabs/Sia/encoding"
+	"github.com/NebulousLabs/ed25519"
 )
 
-// Creates and encodes a public key, and verifies that it decodes correctly,
-// does the same with a signature.
-func TestSignatureEncoding(t *testing.T) {
+// mockKeyDeriver is a mock implementation of keyDeriver that saves its provided
+// entropy and allows the client to specify the returned SecretKey and
+// PublicKey.
+type mockKeyDeriver struct {
+	called  bool
+	entropy [EntropySize]byte
+	sk      ed25519.SecretKey
+	pk      ed25519.PublicKey
+}
+
+func (kd *mockKeyDeriver) deriveKeyPair(entropy [EntropySize]byte) (ed25519.SecretKey, ed25519.PublicKey) {
+	kd.called = true
+	kd.entropy = entropy
+	return kd.sk, kd.pk
+}
+
+// Test that the Generate method is properly calling its dependencies and
+// returning the expected key pair.
+func TestGenerateRandomKeyPair(t *testing.T) {
+	var mockEntropy [EntropySize]byte
+	mockEntropy[0] = 5
+	mockEntropy[EntropySize-1] = 5
+	entropyReader := bytes.NewReader(mockEntropy[:])
+
+	sk := ed25519.SecretKey(&[SecretKeySize]byte{})
+	sk[0] = 7
+	sk[32] = 8
+	pk := ed25519.PublicKey(&[PublicKeySize]byte{})
+	pk[0] = sk[32]
+	kd := mockKeyDeriver{sk: sk, pk: pk}
+
+	// Create a SignatureKeyGenerator using mocks.
+	g := stdGenerator{entropyReader, &kd}
+
 	// Create key pair.
-	sk, pk, err := GenerateSignatureKeys()
+	skActual, pkActual, err := g.Generate()
+
+	// Verify that we got back the expected results.
 	if err != nil {
-		t.Fatal(err)
+		t.Error(err)
+	}
+	if *sk != skActual {
+		t.Errorf("Generated secret key does not match expected! expected = %v, actual = %v", sk, skActual)
+	}
+	if *pk != pkActual {
+		t.Errorf("Generated public key does not match expected! expected = %v, actual = %v", pk, pkActual)
+	}
+
+	// Verify the dependencies were called correctly
+	if !kd.called {
+		t.Error("keyDeriver was never called.")
+	}
+	if mockEntropy != kd.entropy {
+		t.Errorf("keyDeriver was called with the wrong entropy. expected = %v, actual = %v", mockEntropy, kd.entropy)
+	}
+}
+
+// failingReader is a mock implementation of io.Reader that fails with a client-
+// defined error.
+type failingReader struct {
+	err error
+}
+
+func (fr failingReader) Read([]byte) (int, error) {
+	return 0, fr.err
+}
+
+// Test that the Generate method fails if the call to entropy source fails
+func TestGenerateRandomKeyPairFailsWhenRandFails(t *testing.T) {
+	fr := failingReader{err: errors.New("mock error from entropy reader")}
+	g := stdGenerator{entropySource: &fr}
+	if _, _, err := g.Generate(); err == nil {
+		t.Error("Generate should fail when entropy source fails.")
+	}
+}
```

Comment:
I think we're all agreed that it's not necessary in this PR, but it is something I think is good to come to agreement on in general.

I'm *pretty* sure we're in agreement that returning `nil` values when the `error` is non-`nil` is good practice, but not part of the function's contract.

It seems like @DavidVorick is saying this kind of testing is a "nice to have, but not necessary" but I think it's actually counterproductive to do this kind of testing and we should avoid it in our tests. Testing implementation details violates the TDD principle of ["test the interface, not the behavior"](http://googletesting.blogspot.com/2013/08/testing-on-toilet-test-behavior-not.html). A failing unit test should be a sign to the developer that they're changing the class/function's behavior, so I recommend avoiding practices that add noise to this signal.